### ISSUE TITLE
Encode output does not match 1-best from NBestEncode

### ISSUE THREAD

**Issue by nikitakit** [google/sentencepiece#747]
I just came across a case where the tokenization from the Encode family of methods doesn't match the 1-best tokenization from the NBestEncode methods. My understanding is that both outputs have identical scores under the unigram model, and that this comes down to different tie-breaking rules when selecting a single output to return. It still seems potentially undesirable, because this can result in situations where different codepaths settle on different conventions (e.g. a potential training vs. inference mismatch)

Repro:

```python
# Download T5 tokenizer
!wget https://huggingface.co/t5-base/resolve/main/spiece.model

# The assertion at the end will fail
import sentencepiece as spm
tokenizer = spm.SentencePieceProcessor(model_file='spiece.model')
text = 'test.........'
print(tokenizer.EncodeAsPieces(text))  # ['▁test', '........', '.']
print(tokenizer.NBestEncodeAsPieces(text, 1)[0]) # ['▁test', '.', '........']
assert tokenizer.EncodeAsPieces(text) == tokenizer.NBestEncodeAsPieces(text, 1)[0]

**Comment by taku910** [comment]
1-best decoding uses an efficient implementation that doesn't build a lattice internally, which may cause different tokenization.
I presume that they produce the same segmentation but it is not 100% guaranteed when the score is tie.

https://github.com/google/sentencepiece/blob/master/src/unigram_model.cc#L301  (Nbest)
https://github.com/google/sentencepiece/blob/master/src/unigram_model.cc#L815 (1-best, optimized)

We will change the code to call the optimized implementation when N=1.

**Comment by taku910** [comment]
Fixed this issue in version 0.1.97.
